scipy/optimize/_numdiff.py

```diff
@@ -0,0 +1,320 @@
+        applies to center schemes.
+    lb : None or array_like
+        Lower bounds on independent variables, None means no bounds.
+    ub : None or array_like
+        Upper bounds on independent variables, None means no bounds.
+
+    Returns
+    -------
+    h_adjusted : ndarray
+        Adjusted step sizes. Step sizes decrease only if sign flip or
+        switching to one-sided scheme still doesn't allow to take full
+        steps.
+    use_one_sided : ndarray of bool
+        Whether to switch to one-sided scheme. Informative only for
+        `scheme` = '2-sided'.
+    """
+    if scheme == '1-sided':
+        use_one_sided = np.ones_like(h, dtype=bool)
+    elif scheme == '2-sided':
+        h = np.abs(h)
+        use_one_sided = np.zeros_like(h, dtype=bool)
+    else:
+        raise ValueError("`scheme` must be '1-sided' or '2-sided'.")
+
+    if lb is None and ub is None:
+        return h, use_one_sided
+
+    if lb is None:
+        lb = np.empty_like(h)
+        lb.fill(-np.inf)
+
+    if ub is None:
+        ub = np.empty_like(h)
+        ub.fill(np.inf)
+
+    h_total = h * num_steps
+    h_adjusted = h.copy()
+
+    lower_dist = x0 - lb
+    upper_dist = ub - x0
+
+    if scheme == '1-sided':
+        x = x0 + h_total
+        violated = (x < lb) | (x > ub)
+        fitting = np.abs(h_total) <= np.maximum(lower_dist, upper_dist)
+        h_adjusted[violated & fitting] *= -1
+
+        forward = (upper_dist >= lower_dist) & ~fitting
+        h_adjusted[forward] = upper_dist[forward] / num_steps
+        backward = (upper_dist < lower_dist) & ~fitting
+        h_adjusted[backward] = -lower_dist[backward] / num_steps
+    elif scheme == '2-sided':
+        central = (lower_dist >= h_total) & (upper_dist >= h_total)
+
+        forward = (upper_dist >= lower_dist) & ~central
+        h_adjusted[forward] = np.minimum(
+            h[forward], 0.5 * upper_dist[forward] / num_steps)
+        use_one_sided[forward] = True
+
+        backward = (upper_dist < lower_dist) & ~central
+        h_adjusted[backward] = -np.minimum(
+            h[backward], 0.5 * lower_dist[backward] / num_steps)
+        use_one_sided[backward] = True
+
+        min_dist = np.minimum(upper_dist, lower_dist) / num_steps
+        adjusted_central = (~central & (np.abs(h_adjusted) <= min_dist))
+        h_adjusted[adjusted_central] = min_dist[adjusted_central]
+        use_one_sided[adjusted_central] = False
+
+    return h_adjusted, use_one_sided
+
+
+def _compute_absolute_step(rel_step, x0, method):
+    if rel_step is None:
+        if method == '2-point':
+            rel_step = EPS**0.5
+        elif method == '3-point':
+            rel_step = EPS**(1 / 3)
+        else:
+            raise ValueError("`method` must be '2-point' or '3-point'.")
+
+    sign_x0 = (x0 >= 0).astype(float) * 2 - 1
+    return np.atleast_1d(rel_step * sign_x0 * np.maximum(1.0, np.abs(x0)))
+
+
+def approx_derivative(f, x0, method='3-point', rel_step=None, f0=None,
```

Comment:
Please do not enforce a particular parallelization mechanism within the function. The method emcee uses is what I described: it sets up a batch of function evaluations, then evaluates them with map() - but it can be a version of map() that the user provides as an argument, so that the user can supply a version of map() that uses multiprocessing, or joblib, or MPI, or some custom hack.